File: sdk/dart/lib/src/services.dart
import 'http_client.dart';

/// Services: list, details, cancel, renew, billing cycle
class ServicesResource {
  final HttpClient httpClient;
  ServicesResource(this.httpClient);

  /// GET /api/service - List all services
  Future<Map<String, dynamic>> list([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service', params);
  }

  /// GET /api/service/{id} - Return service details
  Future<Map<String, dynamic>> get([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}', params);
  }

  /// GET /api/service/{id}/methods - List methods available for service
  Future<Map<String, dynamic>> listMethods([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/methods', params);
  }

  /// POST /api/service/{id}/cancel - Request service cancellation
  Future<Map<String, dynamic>> cancel([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/cancel', params);
  }

  /// GET /api/service/{id}/label - Show current service label
  Future<Map<String, dynamic>> getLabel([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/label', params);
  }

  /// POST /api/service/{id}/label - Set new custom label
  Future<Map<String, dynamic>> setLabel([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/label', params);
  }

  /// POST /api/service/{id}/renew - Generate renewal invoice
  Future<Map<String, dynamic>> renew([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/renew', params);
  }

  /// GET /api/service/{id}/cycle - Get billing cycle options
  Future<Map<String, dynamic>> listBillingCycles([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/cycle', params);
  }

  /// POST /api/service/{id}/cycle - Change billing cycle
  Future<Map<String, dynamic>> changeBillingCycle([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/cycle', params);
  }

}

File: sdk/julia/src/services.jl
# Services: list, details, cancel, renew, billing cycle
struct ServicesResource
    http::HttpClient
end

# GET /api/service - List all services
function list(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service"; params = params)
end

# GET /api/service/{id} - Return service details
function get(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}"; params = params)
end

# GET /api/service/{id}/methods - List methods available for service
function list_methods(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/methods"; params = params)
end

# POST /api/service/{id}/cancel - Request service cancellation
function cancel(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/cancel"; params = params)
end

# GET /api/service/{id}/label - Show current service label
function get_label(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/label"; params = params)
end

# POST /api/service/{id}/label - Set new custom label
function set_label(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/label"; params = params)
end

# POST /api/service/{id}/renew - Generate renewal invoice
function renew(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/renew"; params = params)
end

# GET /api/service/{id}/cycle - Get billing cycle options
function list_billing_cycles(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/cycle"; params = params)
end

# POST /api/service/{id}/cycle - Change billing cycle
function change_billing_cycle(r::ServicesResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/cycle"; params = params)
end
